### ISSUE TITLE
testStopReplicatorAfterOffline_SG fails on iOS 10.2 simulator

### ISSUE THREAD

**Issue by pasin** [couchbase/couchbase-lite-ios#2089]
The test passes on 11.x simulators.

BT

```
thread #1, queue = 'com.apple.main-thread'
    frame #0: 0x0000000114f517c2 libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x0000000114f50cdc libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x000000010ddc8ba4 CoreFoundation`__CFRunLoopServiceMachPort + 212
    frame #3: 0x000000010ddc8021 CoreFoundation`__CFRunLoopRun + 1345
    frame #4: 0x000000010ddc7884 CoreFoundation`CFRunLoopRunSpecific + 420
    frame #5: 0x000000010d7b64ad XCTest`-[XCTWaiter waitForExpectations:timeout:enforceOrder:] + 545
    frame #6: 0x000000010d7674d5 XCTest`-[XCTestCase(AsynchronousTesting) _waitForExpectations:timeout:enforceOrder:handler:] + 132
    frame #7: 0x000000012214b6d1 CouchbaseLiteTests`-[ReplicatorTest testStopReplicatorAfterOffline_SG](self=0x000060700002a9c0, _cmd="testStopReplicatorAfterOffline_SG") at ReplicatorTest.m:848
    frame #8: 0x000000010ddc544c CoreFoundation`__invoking___ + 140
    frame #9: 0x000000010ddc52d1 CoreFoundation`-[NSInvocation invoke] + 289
    frame #10: 0x000000010d773e30 XCTest`__24-[XCTestCase invokeTest]_block_invoke + 591
    frame #11: 0x000000010d7bd17e XCTest`-[XCUITestContext performInScope:] + 183
    frame #12: 0x000000010d773bd6 XCTest`-[XCTestCase invokeTest] + 141
    frame #13: 0x000000010d774b97 XCTest`__26-[XCTestCase performTest:]_block_invoke.369 + 42
    frame #14: 0x000000010d7c1f25 XCTest`+[XCTContext runInContextForTestCase:block:] + 163
    frame #15: 0x000000010d774533 XCTest`-[XCTestCase performTest:] + 608
    frame #16: 0x000000010d770539 XCTest`__27-[XCTestSuite performTest:]_block_invoke + 363
    frame #17: 0x000000010d76fea0 XCTest`-[XCTestSuite _performProtectedSectionForTest:testSection:] + 26
    frame #18: 0x000000010d77009d XCTest`-[XCTestSuite performTest:] + 239
    frame #19: 0x000000010d770539 XCTest`__27-[XCTestSuite performTest:]_block_invoke + 363
    frame #20: 0x000000010d76fea0 XCTest`-[XCTestSuite _performProtectedSectionForTest:testSection:] + 26
    frame #21: 0x000000010d77009d XCTest`-[XCTestSuite performTest:] + 239
    frame #22: 0x000000010d770539 XCTest`__27-[XCTestSuite performTest:]_block_invoke + 363
    frame #23: 0x000000010d76fea0 XCTest`-[XCTestSuite _performProtectedSectionForTest:testSection:] + 26
    frame #24: 0x000000010d77009d XCTest`-[XCTestSuite performTest:] + 239
    frame #25: 0x000000010d7c964f XCTest`__44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke + 40
    frame #26: 0x000000010d78371a XCTest`-[XCTestObservationCenter _observeTestExecutionForBlock:] + 475
    frame #27: 0x000000010d7c94ee XCTest`-[XCTTestRunSession runTestsAndReturnError:] + 281
    frame #28: 0x000000010d75faf1 XCTest`-[XCTestDriver runTestsAndReturnError:] + 314
    frame #29: 0x000000010d7c1190 XCTest`_XCTestMain + 619
    frame #30: 0x000000010c362591 xctest`___lldb_unnamed_symbol4$$xctest + 350
    frame #31: 0x0000000114c2a68d libdyld.dylib`start + 1
    frame #32: 0x0000000114c2a68d libdyld.dylib`start + 1
  thread #2, queue = 'com.apple.libdispatch-manager'
    frame #0: 0x0000000114f5bf0a libsystem_kernel.dylib`kevent_qos + 10
    frame #1: 0x0000000114bd5c37 libdispatch.dylib`_dispatch_mgr_wait_for_event + 79
    frame #2: 0x0000000114bcfa81 libdispatch.dylib`_dispatch_mgr_invoke + 248
    frame #3: 0x0000000114bcf883 libdispatch.dylib`_dispatch_mgr_thread + 54
  thread #3
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #4, name = 'LiteCore Timer'
    frame #0: 0x0000000114f5acee libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x0000000114f92662 libsystem_pthread.dylib`_pthread_cond_wait + 732
    frame #2: 0x0000000114784632 libc++.1.dylib`std::__1::condition_variable::__do_timed_wait(std::__1::unique_lock<std::__1::mutex>&, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > >) + 126
    frame #3: 0x0000000123169ba1 CouchbaseLite`std::__1::cv_status std::__1::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000000l> >(this=0x000060d000017318, __lk=0x00000001247a31f0, __d=0x00000001247a6dd0) at __mutex_base:416
    frame #4: 0x000000012315dea7 CouchbaseLite`std::__1::cv_status std::__1::condition_variable::wait_until<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > >(this=0x000060d000017318, __lk=0x00000001247a31f0, __t=0x00000001247a3270) at __mutex_base:384
    frame #5: 0x000000012315bef1 CouchbaseLite`litecore::actor::Timer::Manager::run(this=0x000060d0000172c0) at Timer.cc:73
    frame #6: 0x0000000123164a86 CouchbaseLite`litecore::actor::Timer::Manager::Manager(this=0x0000602000009c18)::$_0::operator()() const at Timer.cc:33
    frame #7: 0x0000000123164618 CouchbaseLite`std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, litecore::actor::Timer::Manager::Manager()::$_0> >(void*, void*) [inlined] std::__1::__invoke<litecore::actor::Timer::Manager::Manager()::$_0>(__f=0x0000602000009c18)::$_0>(fp)(std::__1::forward<>(fp0))), litecore::actor::Timer::Manager::Manager()::$_0&&) at type_traits:4291
    frame #8: 0x00000001231645fe CouchbaseLite`std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, litecore::actor::Timer::Manager::Manager()::$_0> >(void*, void*) [inlined] _ZNSt3__116__thread_executeINS_10unique_ptrINS_15__thread_structENS_14default_deleteIS2_EEEEZN8litecore5actor5Timer7ManagerC1EvE3$_0JEJEEEvRNS_5tupleIJT_T0_DpT1_EEENS_15__tuple_indicesIJXspT2_EEEE(__t=0x0000602000009c10) at thread:336
    frame #9: 0x00000001231645c2 CouchbaseLite`std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, litecore::actor::Timer::Manager::Manager()::$_0> >(__vp=0x0000602000009c10) at thread:346
    frame #10: 0x0000000114f916c1 libsystem_pthread.dylib`_pthread_body + 340
    frame #11: 0x0000000114f9156d libsystem_pthread.dylib`_pthread_start + 377
    frame #12: 0x0000000114f90c5d libsystem_pthread.dylib`thread_start + 13
  thread #5, queue = 'parsing queue'
    frame #0: 0x0000000114f517fe libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x0000000114bcb04a libdispatch.dylib`_dispatch_semaphore_wait_slow + 171
    frame #2: 0x0000000119951060 DTXConnectionServices`-[DTXMessageParser waitForMoreData:incrementalBuffer:] + 87
    frame #3: 0x0000000119950b42 DTXConnectionServices`-[DTXMessageParser parseMessageWithExceptionHandler:] + 64
    frame #4: 0x00000001199508b5 DTXConnectionServices`__68-[DTXMessageParser initWithMessageHandler:andParseExceptionHandler:]_block_invoke + 39
    frame #5: 0x000000010c43e122 libclang_rt.asan_iossim_dynamic.dylib`__wrap_dispatch_async_block_invoke + 290
    frame #6: 0x0000000114bbf808 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #7: 0x0000000114be112e libdispatch.dylib`_dispatch_client_callout + 8
    frame #8: 0x0000000114bc64cf libdispatch.dylib`_dispatch_queue_serial_drain + 1018
    frame #9: 0x0000000114bc6c9f libdispatch.dylib`_dispatch_queue_invoke + 1118
    frame #10: 0x0000000114bc89dc libdispatch.dylib`_dispatch_root_queue_drain + 506
    frame #11: 0x0000000114bc8782 libdispatch.dylib`_dispatch_worker_thread3 + 113
    frame #12: 0x0000000114f911ca libsystem_pthread.dylib`_pthread_wqthread + 1387
    frame #13: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #6
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #7
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #8
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #9
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #10, queue = 'BLIP[->ws:foo.couchbase.com/db/_blipsync]'
    frame #0: 0x000000010c437381 libclang_rt.asan_iossim_dynamic.dylib`__asan_memcpy + 1
    frame #1: 0x0000000122b474d9 CouchbaseLite`litecore::websocket::CloseStatus::operator=(this=0x000060c00008d460, (null)=0x0000000126c9bae0) at WebSocketInterface.hh:73
    frame #2: 0x00000001231ae911 CouchbaseLite`litecore::blip::Connection::closed(this=0x000060c00008d400, status=litecore::blip::Connection::CloseStatus @ 0x0000700008aaef58) at BLIPConnection.cc:699
    frame #3: 0x00000001231bac32 CouchbaseLite`litecore::blip::BLIPIO::_closed(this=0x000061600005e580, status=CloseStatus @ 0x0000700008aaf150) at BLIPConnection.cc:215
    frame #4: 0x00000001231bb0f5 CouchbaseLite`::___ZN8litecore5actor5Actor7enqueueINS_4blip6BLIPIOEJNS_9websocket11CloseStatusEEEEvMT_FvDpT0_ES9__block_invoke(.block_descriptor=<unavailable>) at Actor.hh:96
    frame #5: 0x000000012316f0b2 CouchbaseLite`litecore::actor::safelyCall(block=0x00000001231baeb0) block_pointer) at GCDMailbox.cc:79
    frame #6: 0x000000012316f04e CouchbaseLite`::___ZN8litecore5actor10GCDMailbox7enqueueEU13block_pointerFvvE_block_invoke(.block_descriptor=<unavailable>) at GCDMailbox.cc:95
    frame #7: 0x000000010c43e122 libclang_rt.asan_iossim_dynamic.dylib`__wrap_dispatch_async_block_invoke + 290
    frame #8: 0x0000000114bbf808 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #9: 0x0000000114be112e libdispatch.dylib`_dispatch_client_callout + 8
    frame #10: 0x0000000114bc64cf libdispatch.dylib`_dispatch_queue_serial_drain + 1018
    frame #11: 0x0000000114bc6c9f libdispatch.dylib`_dispatch_queue_invoke + 1118
    frame #12: 0x0000000114bc89dc libdispatch.dylib`_dispatch_root_queue_drain + 506
    frame #13: 0x0000000114bc8782 libdispatch.dylib`_dispatch_worker_thread3 + 113
    frame #14: 0x0000000114f911ca libsystem_pthread.dylib`_pthread_wqthread + 1387
    frame #15: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #11
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #12
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
* thread #13, queue = 'com.apple.NSURLSession-work', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
  * frame #0: 0x00000001141d9757 CFNetwork`-[__NSCFURLLocalStreamTask _onSessionQueue_disavow] + 34
    frame #1: 0x000000011403da97 CFNetwork`__51-[NSURLSession delegate_task:didCompleteWithError:]_block_invoke_2 + 31
    frame #2: 0x000000010c43e122 libclang_rt.asan_iossim_dynamic.dylib`__wrap_dispatch_async_block_invoke + 290
    frame #3: 0x0000000114bbf808 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #4: 0x0000000114be112e libdispatch.dylib`_dispatch_client_callout + 8
    frame #5: 0x0000000114bc64cf libdispatch.dylib`_dispatch_queue_serial_drain + 1018
    frame #6: 0x0000000114bc6c9f libdispatch.dylib`_dispatch_queue_invoke + 1118
    frame #7: 0x0000000114bc89dc libdispatch.dylib`_dispatch_root_queue_drain + 506
    frame #8: 0x0000000114bc8782 libdispatch.dylib`_dispatch_worker_thread3 + 113
    frame #9: 0x0000000114f911ca libsystem_pthread.dylib`_pthread_wqthread + 1387
    frame #10: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #14
    frame #0: 0x0000000114f5b562 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000114f9126f libsystem_pthread.dylib`_pthread_wqthread + 1552
    frame #2: 0x0000000114f90c4d libsystem_pthread.dylib`start_wqthread + 13
  thread #15, name = 'com.apple.CFStream.LegacyThread'
    frame #0: 0x0000000114f517c2 libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x0000000114f50cdc libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x000000010ddc8ba4 CoreFoundation`__CFRunLoopServiceMachPort + 212
    frame #3: 0x000000010ddc8021 CoreFoundation`__CFRunLoopRun + 1345
    frame #4: 0x000000010ddc7884 CoreFoundation`CFRunLoopRunSpecific + 420
    frame #5: 0x000000010de7f3c5 CoreFoundation`_legacyStreamRunLoop_workThread + 261
    frame #6: 0x0000000114f916c1 libsystem_pthread.dylib`_pthread_body + 340
    frame #7: 0x0000000114f9156d libsystem_pthread.dylib`_pthread_start + 377
    frame #8: 0x0000000114f90c5d libsystem_pthread.dylib`thread_start + 13
  thread #16, name = 'com.apple.NSURLConnectionLoader'
    frame #0: 0x0000000114f517c2 libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x0000000114f50cdc libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x000000010ddc8ba4 CoreFoundation`__CFRunLoopServiceMachPort + 212
    frame #3: 0x000000010ddc8021 CoreFoundation`__CFRunLoopRun + 1345
    frame #4: 0x000000010ddc7884 CoreFoundation`CFRunLoopRunSpecific + 420
    frame #5: 0x0000000113fbab98 CFNetwork`+[NSURLConnection(Loader) _resourceLoadLoop:] + 361
    frame #6: 0x000000010d250654 Foundation`__NSThread__start__ + 1243
    frame #7: 0x0000000114f916c1 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x0000000114f9156d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x0000000114f90c5d libsystem_pthread.dylib`thread_start + 13
```

**Comment by pasin** [comment]
This seems to be an issue with NSURLSession's stream task on iOS 10.2 that canceling the task after an error could cause the crash down in `__NSCFURLLocalStreamTask`. I also noticed a weird issue that -URLSession:task:didCompleteWithError: is called twice.

**Comment by snej** [comment]
I've posted a question to Apple's macnetworkprog mailing list.

**Comment by pasin** [comment]
I have did some more experiment on iOS 10.2. Here are what I have found:

1. The crash will not happen if canceling the task when there is no error.
2. The crash will not happen if canceling the task due to OperationTimeout (This is a cause of the socket leak in #2078). The crash will happen with connection related errors such as connection refuse (-1004) which I don't see socket leaks for these errors.
3. The crash will not happen if I put some delay (e.g. 0.3 secs) before canceling. 
4. Put canceling in the operation queue or calling `[_session invalidateAndCancel] makes the issue a little more harder to reproduce but the issue could still happen.

From the 2nd bullet, one thing that we can do for 2.0 is to cancel the task only when receiving OperationTimeout.
